PR for Database changes

wadoodalam opened the pull request:
# Description

This PR modifies the database by making changes to meters, maps, and groups. Additional attributes for meters are added which allows to check if the given reading is the cumulative sum or not and if and when the pipeline needs reset. Other changes include the timestamps, variation, and length of the reading associated with the meter. It also allows for checking if the reading overlaps to the previous day or not. This PR is also a step towards determining the start time of the reading if not provided. Additionally, The PR allows for the storage of some additional data with regards to groups. Another thing addressed in this PR is to allow for storing the circle size of the meter on the map and the angle with which the image has been rotated with respect to the north. Test code for these changes and changes to the script of test data has also been implemented.

Fixes #634. It begins to address issue #211.

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request

## Limitations
Test code associated with OED/src/server/test/web/meters.js  and OED/src/server/test/web/groups.js still needs fixes. Additional attributes which are added in this PR are not returned by the relevant route from the server to the aforementioned files.

by @huss: The test code has been updated and no longer fails. I have the new values so they are returned by the route to allow testing but they have not been added to Redux state at this point. It is assumed that will happen as the code that will use these values are added. As such, this PR updates the DB to hold state OED is planning to use so future developers can add that code.

wadoodalam committed:
Merge pull request #1 from OpenEnergyDashboard/development

Update

wadoodalam committed:
Merge pull request #2 from OpenEnergyDashboard/development

0.6

wadoodalam committed:
Changes to DB regarding meters and maps

wadoodalam committed:
moved to 0.6-0.7

wadoodalam committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into development

wadoodalam committed:
lints check and test check error removed

wadoodalam committed:
Changes to meters, groups and maps.
Array length test for groups not passing

wadoodalam committed:
Maps and Meters test

wadoodalam committed:
meter tests

wadoodalam committed:
Modified further tests. Chnaged the default values for maps

wadoodalam committed:
Tests for web edited

wadoodalam committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into wadoodalam

wadoodalam committed:
updated

wadoodalam committed:
commented code for raw readings

wadoodalam committed:
Merge pull request #6 from OpenEnergyDashboard/development

Update branch

wadoodalam committed:
script fix for test data

wadoodalam commented:
Database changes for meters, maps, and groups. 
Models are modified to accommodate the DB changes. Additionally, the test code and test date script was modified/created(1 test in web/groups.js needs a fix).

huss commented on src/server/migrations/0.6.0-0.7.0/sql/group/add_displayable,gps,note,area_columns.sql:
```diff
@@ -0,0 +1,10 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This file name is not consistent with other OED files. There should not be commas in the name.

huss commented on src/server/migrations/0.6.0-0.7.0/sql/group/add_displayable,gps,note,area_columns.sql:
```diff
@@ -0,0 +1,10 @@
+ALTER TABLE groups
+	ADD COLUMN IF NOT EXISTS displayable BOOLEAN,
+    ADD COLUMN IF NOT EXISTS gps POINT DEFAULT NULL,
+    ADD COLUMN IF NOT EXISTS note VARCHAR(500),
```
From the documentation, VARCHAR(#) and TEXT are internally stored the same way in Postgres. It seems best to use TEXT unless you have a reason to restrict the length of the attribute. Given we have a limited number of groups, meters, etc., I don't see a reason to limit the length of the note. We just have to be careful to allow for longer ones when we display on the admin page. Thus, can you please change this to TEXT. Also, please make the same change with migrations for the note in meters and maps tables. The other uses of VARCHAR in OED are probably reasonable since the actual item should be of limited length so we will not change those.

huss commented on src/server/migrations/0.6.0-0.7.0/sql/group/add_displayable,gps,note,area_columns.sql:
```diff
@@ -0,0 +1,10 @@
+	ADD COLUMN IF NOT EXISTS displayable BOOLEAN,
+    ADD COLUMN IF NOT EXISTS gps POINT DEFAULT NULL,
+    ADD COLUMN IF NOT EXISTS note VARCHAR(500),
+    ADD COLUMN IF NOT EXISTS area FLOAT(8) DEFAULT NULL;
```
FLOAT(#) in most DB systems (including Postgres) is somewhat misleading to use since it generally either goes to single or double precision. OED is not 100% consistent but it tends to use REAL (double precision) for values. It has not used (FLOAT(#). Given the limited number of rows, I see no reason not to use REAL here (note we use REAL in readings and that is a lot of rows).

huss commented on src/server/migrations/0.6.0-0.7.0/sql/map/add_column_north_angle and circe_fraction.sql:
```diff
@@ -0,0 +1,7 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+ALTER TABLE maps
+	ADD COLUMN IF NOT EXISTS north_angle FLOAT(8) DEFAULT 0.0,
```
FLOAT(8)->REAL

huss commented on src/server/migrations/0.6.0-0.7.0/sql/map/add_column_north_angle and circe_fraction.sql:
```diff
@@ -0,0 +1,7 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
File name has spaces and circe is mispelled. Please make it consistent with OED file names.

huss commented on src/server/migrations/0.6.0-0.7.0/sql/meter/add_multiple_db_meters_columns.sql:
```diff
@@ -0,0 +1,19 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+ALTER TABLE meters
+	ADD COLUMN IF NOT EXISTS note VARCHAR(500),
```
varchar→text and float(8)->real.

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -23,5 +35,17 @@ CREATE TABLE IF NOT EXISTS meters (
+	reading_variation TIME DEFAULT '23:59:59',
+	reading REAL DEFAULT 0.0,
+	start_timestamp TIMESTAMP DEFAULT '0001-01-01 : 00:00:00',
+	end_timestamp TIMESTAMP DEFAULT '0001-01-01 : 00:00:00'   
```
extra spaces at end of line.

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -5,14 +5,26 @@
+start_timestamp:		Start timestamp of reading
+end_timestamp:			End timestamp of reading
 */
 CREATE TABLE IF NOT EXISTS meters (
```
I know the original file had tabs but all indent should be spaces in SQL files.

huss commented on src/server/migrations/0.6.0-0.7.0/sql/meter/add_multiple_db_meters_columns.sql:
```diff
@@ -0,0 +1,19 @@
+    ADD COLUMN IF NOT EXISTS cumulative BOOLEAN DEFAULT false,
+    ADD COLUMN IF NOT EXISTS cumulative_reset BOOLEAN DEFAULT false,
+    ADD COLUMN IF NOT EXISTS cumulative_reset_start TIME DEFAULT '00:00:00',
+    ADD COLUMN IF NOT EXISTS cumulative_reset_end TIME DEFAULT '23:59:50',
```
Per current version of design document, the end time should be 23:59:59.999999, If cannot have decimal on then 23:59:59.
Also make sure reading variation has a consistent default value.

huss commented on src/server/models/Map.js:
```diff
@@ -17,8 +17,10 @@ class Map {
 	 * @param opposite {Point} coordinates of opposite corner from origin
 	 * @param mapSource data URL of image of the map
+	 * @param northAngle stores angle between map orientation and true north
+	 * @param maxCircleSizeFraction Stores the fraction of horizontle map
```
horizontal sp.

huss commented on src/server/models/Meter.js:
```diff
@@ -18,8 +18,22 @@ class Meter {
+	 * @param previousDay Checks if the if the reading is of the previous day
+	 * @param readingLength Specifies the time range on every reading in the CSV file
+	 * @param readingVariation +/- time allowed on length to consider within allowed length
+	 * @param reading The value of reading
```
reading, startTimestemp & endTimestamp are the last reading input for this meter into OED. The comment should reflect this.

huss commented on src/server/models/Meter.js:
```diff
@@ -18,8 +18,22 @@ class Meter {
 	 * @param identifier Another way to identify a meter
+	 * @param note Note about the meter
+	 * @param area Area of the meter 
+	 * @param cumulative Checks if the reading is a sum of usage or the particular reading
```
This should be "Identifies if meter readings that are stored are the sum of usage...."

For this and following values, I think it would be good to identify which ones will default to a value (via the DB) if no value is given (does that mean unknown or null?).

huss commented on src/server/models/Meter.js:
```diff
@@ -18,8 +18,22 @@ class Meter {
+	 * @param cumulativeReset Tells if the system reset is needed or not
+	 * @param cumulativeResetStart The earliest time of day that a reset can occur
+	 * @param cumulativeResetEnd The latest time of day that a reset can occur
+	 * @param previousDay Checks if the if the reading is of the previous day
```
The wording of this comment is off but that does not really matter since the latest design document does not have this value.

huss commented on src/server/models/Map.js:
```diff
@@ -17,8 +17,10 @@ class Map {
 	 * @param origin {Point} coordinates of (0,0) on map
 	 * @param opposite {Point} coordinates of opposite corner from origin
 	 * @param mapSource data URL of image of the map
+	 * @param northAngle stores angle between map orientation and true north
```
For this and following value, I think it would be good to identify which ones will default to a value (via the DB) if no value is given (does that mean unknown or null?).

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -5,14 +5,26 @@
+note:					Notes about the meter
+area:					Area covered by the meter
+cumulative:				True if readings provided are the sum usage and not the value for this particular reading
+cumulative_reset:		True if pipline is to be reset
```
I think a better comment is "True if cumulative values can reset back to zero."

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -5,14 +5,26 @@
+previous_day: 			True if reading is from previous day 
+reading_length:			Specify the time range on every reading
+reading_variation: 		+/- time allowed on length to consider within allowed length
+reading: 				The reading from the meter
```
Make sure these comment and other items from migration get into this file.

huss commented on src/server/sql/meter/get_meter_by_id.sql:
```diff
@@ -2,5 +2,5 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-SELECT id, name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier FROM meters
+SELECT id, name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier, note, area, cumulative, cumulative_reset, cumulative_reset_start, cumulative_reset_end, previous_day, reading_length, reading_variation, reading, start_timestamp, end_timestamp FROM meters
```
This line is too long. Create multiple lines.

huss commented on src/server/sql/meter/get_meter_by_name.sql:
```diff
@@ -2,5 +2,5 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-SELECT id, name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier FROM meters
+SELECT id, name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier, note, area, cumulative, cumulative_reset, cumulative_reset_start, cumulative_reset_end, previous_day, reading_length, reading_variation, reading, start_timestamp, end_timestamp FROM meters
```
This line is too long. Create multiple lines.

huss commented on src/server/sql/meter/insert_new_meter.sql:
```diff
@@ -2,9 +2,6 @@
-*/
-INSERT INTO meters(name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier)
-    VALUES (${name}, ${ipAddress}, ${enabled}, ${displayable}, ${type}, ${meterTimezone}, ${gps}, ${identifier})
+INSERT INTO meters(name, ipaddress, enabled, displayable, meter_type, default_timezone_meter, gps, identifier, note, area, cumulative, cumulative_reset, cumulative_reset_start,cumulative_reset_end, previous_day, reading_length, reading_variation, reading, start_timestamp, end_timestamp)
```
These lines are too long. Create multiple lines.

huss commented on src/server/sql/reading/create_compressed_reading_views.sql:
```diff
@@ -76,6 +76,7 @@ minutely_readings
 						'1 minute'::INTERVAL
 				) gen(interval_start)
 			GROUP BY r.meter_id, gen.interval_start;
+			
```
No tabs on blank lines.

huss commented on src/server/sql/reading/create_compressed_reading_views.sql:
```diff
@@ -193,17 +194,24 @@ BEGIN
 			INNER JOIN unnest(meter_ids) meters(id) ON hourly.meter_id = meters.id
 		WHERE requested_range @> time_interval;
-	ELSE
+	/* Change the code below and get meterid, readings, start and end timestamp from readings table instead(raw data)*/
```
These are changes outside this PR so the comments need to be removed.

huss commented on src/server/test/db/groupTests.js:
```diff
@@ -11,24 +11,36 @@ const Point = require('../../models/Point');
-	const groupA = new Group(undefined, 'A');
-	const groupB = new Group(undefined, 'B');
-	const groupC = new Group(undefined, 'C');
+	const groupA = new Group(undefined, 'A', true, gps, 'notes', 33.5);
```
While it really isn't important, if it is group A, B, C then I think it should be notes A, notes B, notes C.

huss commented on src/server/test/db/groupTests.js:
```diff
@@ -11,24 +11,36 @@ const Point = require('../../models/Point');
-	const meterB = new Meter(undefined, 'B', null, false, true, Meter.type.MAMAC, null, gps);
-	const meterC = new Meter(undefined, 'C', null, false, true, Meter.type.METASYS, null, gps);
+	const meterA = new Meter(undefined, 'A', null, false, true, Meter.type.MAMAC, null, gps, 
+	'Identified 1' ,'Notes', 35.0, true, true, '01:01:25' , '00:00:00', true, '05:00:00','00:00:00', 1.5,
```
First, in this and the following three meter creation you are inconsistent in identified name. Also, please make them consistent with meter name per group comment above. Second, make note names consistent in same way. Third, the timestamp has a colon between the data and time. I think that is not the normal syntax. Fourth, the start_timestamp has invalid dates (year and day).

huss commented on src/server/test/db/mapTests.js:
```diff
@@ -35,12 +37,12 @@ mocha.describe('Maps', () => {
 		const origin = new Point(0.000001, 0.000001);
 		const opposite = new Point(100.000001, 100.000001);
-		const mapPreInsert = new Map(undefined, 'Map', false, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder');
+		const mapPreInsert = new Map(undefined, 'Map', false, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder', 22.5, 10.5);
```
The fraction should be 0-1. While it does not change the test, I think it is best to use valid values unless testing otherwise.

huss commented on src/server/test/db/meterTests.js:
```diff
@@ -24,12 +24,24 @@ function expectMetersToBeEquivalent(expected, actual) {
 	mocha.it('can be saved and retrieved', async () => {
 		const conn = testDB.getConnection();
-		const meterPreInsert = new Meter(undefined, 'Meter', null, false, true, Meter.type.MAMAC, 'UTC', gps);
+		const meterPreInsert = new Meter(undefined, 'Meter', null, false, true, Meter.type.MAMAC, 'UTC',
```
Please see groupTests.js for comments on dates that impact several places in this file.

huss commented on src/server/test/db/meterTests.js:
```diff
@@ -24,12 +24,24 @@ function expectMetersToBeEquivalent(expected, actual) {
+	expect(actual).to.have.property('previousDay', expected.previousDay);
+	expect(actual).to.have.property('readingLength', expected.readingLength);
+	expect(actual).to.have.property('readingVariation', expected.readingVariation);
 }
```
I'm unclear on why the last three attributes of the meters table involving reading are not verified.

huss commented on src/server/test/web/groups.js:
```diff
@@ -30,25 +30,33 @@ mocha.describe('groups API', () => {
-		groupA = new Group(undefined, 'A');
-		groupB = new Group(undefined, 'B');
-		groupC = new Group(undefined, 'C');
+		groupA = new Group(undefined, 'A', true, gps, 'notes', 33.5);
```
Please see comment in groupTests.js about Group and Meter values that apply to this file.

huss commented on src/server/test/web/groups.js:
```diff
@@ -30,25 +30,33 @@ mocha.describe('groups API', () => {
 		groupC.adoptMeter(meterC.id, conn)]);
 	});
 	mocha.describe('retrieval', () => {
+		// For the test below: AssertionError: expected [ Array(3) ] to be a superset of [ Array(3) ]
```
Please add the PR description under limitations that this test is failing.

huss commented on src/server/test/web/maps.js:
```diff
@@ -35,6 +35,8 @@ function expectMapsToBeEquivalent(maps, length) {
 		expect(map).to.have.property('opposite');
 		expectPointsToBeEquivalent(map.opposite, opposite);
 		expect(map).to.have.property('mapSource', 'placeholder');
+		expect(map).to.have.property('northAngle');
```
I think it is worth making sure the value is what is expected for both these new properties.

huss commented on src/server/test/web/maps.js:
```diff
@@ -52,10 +54,10 @@ mocha.describe('maps API', () => {
-		await new Map(undefined, 'Map 2', true, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder').insert(conn);
-		await new Map(undefined, 'Map 3', true, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder').insert(conn);
-		await new Map(undefined, 'Not Visible', false, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder').insert(conn);
+		await new Map(undefined, 'Map 1', true, null, 'default', moment('2000-10-10'), origin, opposite, 'placeholder', 0.0, 0.0).insert(conn);
```
I think at least one of the tests should have maps where the new values are not just 0.0.

huss commented on src/server/test/web/meters.js:
```diff
@@ -22,10 +22,18 @@ mocha.describe('meters API', () => {
-		await new Meter(undefined, 'Meter 2', '1.1.1.1', true, true, Meter.type.MAMAC, 'TZ2', gps).insert(conn);
-		await new Meter(undefined, 'Meter 3', '1.1.1.1', true, true, Meter.type.MAMAC, 'TZ3', gps).insert(conn);
-		await new Meter(undefined, 'Not Visible', '1.1.1.1', true, false, Meter.type.MAMAC, 'TZ4', gps).insert(conn);
+		await new Meter(undefined, 'Meter 1', '1.1.1.1', true, true, Meter.type.MAMAC, 'TZ1', gps, 
```
See groupTests.js for issues on Meter creation values that apply here.

huss commented on src/server/test/web/meters.js:
```diff
@@ -22,10 +22,18 @@ mocha.describe('meters API', () => {
+		'0011-05-022 : 23:59:59', '2020-07-02 : 01:00:10').insert(conn);
+		await new Meter(undefined, 'Not Visible', '1.1.1.1', true, false, Meter.type.MAMAC, 'TZ4', gps, 
+		'Identified 3' ,'Notes', 35.0, true, true, '01:01:25' , '00:00:00', true, '05:00:00','00:00:00', 1.5, 
+		'0011-05-022 : 23:59:59', '2020-07-02 : 01:00:10').insert(conn);
```
I cannot comment on lines you did not change so I'm putting this comment here. I think you really should be testing the new meter values. You added them to the meters you created but you never verify they have the correct values as the previous values did. I'm also uncertain why some meter have the new values and others were not changed.

huss reviewed (changes requested):
There are a fair number of issues that need to be carefully addressed in this PR. I have not yet tested the code functions properly but wanted to get these comments to the author.

wadoodalam committed:
Errors and issued addressed for PR

huss committed:
fix up start/end time

This is an arbitrary time so I think it best to best at midnight.
Fixed date format.

huss committed:
remove previousDay/previous_day

huss committed:
fix comments

huss committed:
fix comments, date format, spacing

huss committed:
fix fomatting

huss committed:
file missed in staging

huss committed:
uniqueness of meters/groups

Made group and meter names, notes different
Fixed one space issue
Made enter identifier names consistent

huss committed:
put back year format

I messed up year in my edits and am putting back

huss committed:
fix for no previous day

Also made time tests use Meter object used to create meter to do tests.

huss committed:
fix up for removing previous day

huss committed:
fix map tests

- The route for maps needed to return the two new values.
- A previous bug stopped test loop from executing. This likely is why
broken tests were not noticed.
- The modifiedDate test was wrong and fixed.
- The new tests for the new properties was wrong.
- Made new values unique to each map so test was stronger.

huss committed:
fix up for previousDay removal

huss committed:
upgrade meter testing & route new items

- The new DB items for meters were added to what the route returns.
Only area is being returned for non-admin to see.
- Refactored code to make testing of meters easier.
- Added tests for all new DB values.
- Make values in each meter more unique for stronger testing.

huss committed:
Fix failing group test

huss committed:
fix group tests

Takes into account removing previous day and renaming of meters/groups.

huss committed:
remove Mac file added in PR

huss commented:
I have pushed a number of commits to the PR to address:
- A few issues not completed by @wadoodalam that are in the review including removal of previous day per design document
- Remove .DS_Store Mac file that should not be in repo (seems added via this PR) and caused check failure
- Improved/fixed some existing testing and add tests for some new DB items in the PR. This includes fixing the test that failed and noted in the PR description (now edited)
- I note that there are a lot of instances of creating meters, groups & maps that do not pass the new arguments. I have created issue #635 so it can be addressed at some point.

huss commented:
I have completed local testing and it seems to run fun. I believe this is ready to merge.
